### PR TITLE
Feature/resource management

### PR DESCRIPTION
**Pull request by yalania**
Mega refactor for ResourceManagement:

- Ensure thread safe
- Create Resource class
- Create ModuleResourceManagement
- Create common resource cache for all resources
- Hide importers from the rest of the engine

### CODE REVIEW

**Review by mariofv — COMMENTED** [review]
EDIT: Ignore this comment ;D

File: Engine/Module/ModuleResourceManager.cpp [hunk 1]
```diff
@@ -95,6 +95,7 @@ std::pair<bool, std::string> ModuleResourceManager::InternalImport(const File& f
 	}
 	if (file.file_type == FileType::TEXTURE)
 	{
+		std::lock_guard<std::mutex> lock(thread_comunication.thread_mutex);
```
Comment (mariofv) [inline Engine/Module/ModuleResourceManager.cpp:98]: What is being changed exactly in this commit?

File: Engine/Module/ModuleFileSystem.cpp [hunk 2]
```diff
@@ -204,11 +204,14 @@ void ModuleFileSystem::GetAllFilesInPath(const std::string & path, std::vector<s
 	char **i;
```
Comment (mariofv) [inline Engine/Module/ModuleFileSystem.cpp]: Would it be possible to rename this variable to a more descriptive name?

**Review by mariofv — COMMENTED** [review]
Great job! Can you please explain when you have a moment how meta files and Loader works? I really like the template implementation :)

File: Engine/Module/ModuleResourceManager.cpp (hunk 1 above)
Comment (yalania) [reply Engine/Module/ModuleResourceManager.cpp:98]: DevIL is not thread safe and crash if you try to import multiple things at the same time. That is why I added a mutex

**Review by Dartemiss — APPROVED** [review]
Amazing job! All seems correct :D

**Review by mariofv — APPROVED** [review]

**Pull request merged**